gshekhawat0820/EchoHub
# Add image picker prototype

## Description

1. Moves the `GoogleView` into a standalone file for better organization.
2. Adds a button to the `GoogleView` to set an image. The image can either come from the camera or photo library.
3. Implements the `ImagePicker` and `ImagePickerView` files, primarily based on [two](https://www.youtube.com/watch?v=W60nnRFUGaI) [videos](https://www.youtube.com/watch?v=Y-65T0YBOm4) that I watched.
4. Updates `Info.plist` to request permission to access the camera.
5. Updates `EchoHub.xcodeproj/project.pbxproj` with a development team linked to my iCloud account. I'm not really sure what a development team is but I don't think this will cause any problems. This was necessary to connect to my physical device.

## Testing

Testing the camera will require connecting your physical device to your Mac since the simulator does not have a camera. I'd recommend following [this tutorial](https://www.twilio.com/en-us/blog/how-to-test-your-ios-application-on-a-real-device-html).

1. Tap on the Google view.
2. Tap on the "Set image" button. A popup menu will appear.
3. Tap on the "Camera" button.
6. Take a photo. After confirming, the photo will show up in the app.

Try other cases such as taking a photo and clicking the cancel button or sliding the modal away from the top. Try similar cases with the photo library.

## Notes

There is an older method of building iOS applications called UIKit. We are using a newer method called SwiftUI. From my understanding, UIKit provided a rather straightforward way to interact with the camera and photo library. SwiftUI does not, and requires porting over that feature.

It turns out that picking from the photo library and using the camera are implemented in essentially the same way. In `ImagePicker`, there is a function used to specify certain properties of how we want to pick an image:

```swift
func makeUIViewController(context: Context) -> UIImagePickerController {
    let picker = UIImagePickerController();
    picker.sourceType = self.sourceType;
    picker.delegate = context.coordinator;
    return picker;
}
```

Notably, the source type can either be a camera (`UIImagePickerController.SourceType.camera`) or photo library (`UIImagePickerController.SourceType.photoLibrary`). However, it should be noted that the photo library source type is [considered deprecated](https://developer.apple.com/documentation/uikit/uiimagepickercontroller/sourcetype/photolibrary). It may be beneficial to explore the [suggested alternative](https://developer.apple.com/documentation/photokit/phpickerviewcontroller).

Note that navigating away from the screen will reset the image.

## Further steps

- Potentially replace the current photo library implementation with `PHPickerViewController`.
- Move image picker interface to a more appropriate place.
- Save the picked photo.

## Tickets

This is related to two Trello tickets: 

- <https://trello.com/c/ZQ1DDeCW/4-take-picture-using-phone-camera>
- <https://trello.com/c/6gLld9ts/3-pick-photo-from-photo-gallery-to-upload-to-icon>
